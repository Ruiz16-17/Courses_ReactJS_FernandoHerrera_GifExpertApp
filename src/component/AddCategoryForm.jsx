import PropTypes from 'prop-types';
import React, { useState } from 'react';

export const AddCategoryForm = ({setCategories}) => {

    const [inputValue, setInputValue] = useState('');

    const handleInputChange = (event) => {
        setInputValue(event.target.value)
    }

    const handleSubmit = (event) => {

        event.preventDefault();

        if(inputValue.trim().length > 2){
            setCategories(categories => [inputValue,...categories]);
            setInputValue('');
        }
        
    }

    return (

        <form onSubmit={handleSubmit}>

            <input
                type="text"
                value={inputValue}
                onChange={handleInputChange}
            />
        </form>

    )
}

AddCategoryForm.propType = {
    setCategories : PropTypes.func.isRequired
}