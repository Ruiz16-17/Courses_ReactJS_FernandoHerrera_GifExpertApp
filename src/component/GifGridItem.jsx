import PropTypes from 'prop-types';
import React from 'react';

export const GifGridItem = ({id, url, title}) => {
    return (
        <div 
            key={id}
            className="card animate__animated animate__bounce"
        >
            <img src={url} alt={title} />
            <p>{title}</p>
        </div>
    )
}

GifGridItem.propType = {
    id : PropTypes.string.isRequired,
    title : PropTypes.string.isRequired,
    url :PropTypes.string.isRequired
    
}