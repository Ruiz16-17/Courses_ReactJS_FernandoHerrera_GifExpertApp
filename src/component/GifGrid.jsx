import PropTypes from 'prop-types';
import React from 'react';
import { useFetchGif } from '../hooks/useFetchGif';
import { GifGridItem } from './GifGridItem';

export const GifGrid = ({ category }) => {

    const {loading, data: images} = useFetchGif(category);

    return (
        <>

            <h3 className="animate__animated animate__flash">{category}</h3>

            {loading && <p>Loading...</p> }

            <div className="card-grid">
            
                {images.map(image => {

                    return (
                        <GifGridItem
                            key={image.id}
                            {...image}

                        />
                    )
                })}

            </div>
        </>
    )
}

GifGrid.propType = {
    category : PropTypes.string.isRequired
    
}