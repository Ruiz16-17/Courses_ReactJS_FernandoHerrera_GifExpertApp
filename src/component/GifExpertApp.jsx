import React, { useState } from 'react';
import { AddCategoryForm } from './AddCategoryForm';
import { GifGrid } from './GifGrid';

export const GifExpertApp = ({defaultCategories = []}) => {

    const [categories, setCategories] = useState(defaultCategories);

    return (
        <>
            <h2>GifExpertApp</h2>
            
            <AddCategoryForm setCategories={setCategories}/>
            
            <hr/>

            <ol>
                {
                    categories.map(
                        category => {
                            return <GifGrid 
                                key={category}
                                category={category} />
                        }
                    )
                
                }
            </ol>

        </>
    )
}
