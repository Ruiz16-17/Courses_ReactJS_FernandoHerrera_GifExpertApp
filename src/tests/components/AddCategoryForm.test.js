import '@testing-library/jest-dom';
import { shallow } from "enzyme";
import React from 'react';
import { AddCategoryForm } from "../../component/AddCategoryForm";

describe('Test on <AddCategoryForm />', () => {

    const setCategories = jest.fn();

    let wrapper = shallow(<AddCategoryForm setCategories={setCategories}/>); 
    
    beforeEach(() =>{
        jest.clearAllMocks();
        wrapper = shallow(<AddCategoryForm setCategories={setCategories}/>); 
    })

    test('Should show correctly', () => {
        expect(wrapper).toMatchSnapshot();
    })
    
    test('should change the textbox', () => {
        const input = wrapper.find('input');
        const value = 'Hola Mundo';
        input.simulate('change', {target:{value}});
    })
    
    test('not should post the info on submit', () => {
        wrapper.find('form').simulate('submit', {preventDefault(){}});
        expect(setCategories).not.toHaveBeenCalled();
    })
    
    test('should called setCategories and clear the textBox', () => {
        
        const value = 'Hola Mundo';

        const input = wrapper.find('input').simulate('change', {target:{value}});
        
        wrapper.find('form').simulate('submit', {preventDefault(){}});

        expect(setCategories).toHaveBeenCalled();
        expect(setCategories).toHaveBeenCalledWith(expect.any(Function));
        //expect(setCategories).toHaveBeenCalledTimes(2); For a specific number of calls

        expect(input.prop('value')).toBe('');

    })
    
});
