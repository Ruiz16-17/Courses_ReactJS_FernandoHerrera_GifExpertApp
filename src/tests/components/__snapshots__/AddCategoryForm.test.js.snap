// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on <AddCategoryForm /> Should show correctly 1`] = `
<form
  onSubmit={[Function]}
>
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
</form>
`;
